usecases/schema/cache.go

```diff
@@ -163,3 +170,103 @@ func (s *schemaCache) deleteClassState(name string) {
 	defer s.Unlock()
 	delete(s.ShardingState, name)
 }
+
+func (s *schemaCache) unsafeFindClassIf(pred func(*models.Class) bool) *models.Class {
+	for _, cls := range s.ObjectSchema.Classes {
+		if pred(cls) {
+			return cls
+		}
+	}
+	return nil
+}
+
+func (s *schemaCache) unsafeFindClass(className string) *models.Class {
+	for _, c := range s.ObjectSchema.Classes {
+		if c.Class == className {
+			return c
+		}
+	}
+	return nil
+}
+
+func (s *schemaCache) addClass(c *models.Class, ss *sharding.State) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.ShardingState[c.Class] = ss
+	s.ObjectSchema.Classes = append(s.ObjectSchema.Classes, c)
+}
+
+func (s *schemaCache) updateClass(u *models.Class, ss *sharding.State) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if c := s.unsafeFindClass(u.Class); c != nil {
+		*c = *u
+	} else {
+		return errClassNotFound
+	}
+	if ss != nil {
+		s.ShardingState[u.Class] = ss
+	}
+	return nil
+}
+
+func (s *schemaCache) addProperty(class string, p *models.Property) (models.Class, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	c := s.unsafeFindClass(class)
+	if c == nil {
+		return models.Class{}, errClassNotFound
+	}
+
+	// update all at once to prevent race condition with concurrent readers
+	src := c.Properties
+	dest := make([]*models.Property, len(src)+1)
+	copy(dest, src)
+	dest[len(src)] = p
+	c.Properties = dest
+	return *c, nil
+}
+
+// readOnlySchema returns a read only schema
+// Changing the schema outside this package might lead to undefined behavior.
+func (s *schemaCache) readOnlySchema() *models.Schema {
+	s.RLock()
+	defer s.RUnlock()
+	return shallowCopySchema(s.ObjectSchema)
+}
+
+func (s *schemaCache) readOnlyClass(name string) (*models.Class, error) {
```

Comment:
This change totally makes sense, but I'm worrying a bit about performance implications.

I did a similar change at the end of last year and if I remember correctly the locking when getting the class slowed things down considerably. Think of people adding a bunch of tenants while doing other operations.

One way of solving this could be to have a read-only copy cached that is only updated either on update or in a given time-interval and accessing this has its own lock

Reply:
> This change totally makes sense, but I'm worrying a bit about performance implications.

 `readOnlyClass` function returns a shallow copy of the schema. The performance cost is quite negligible because we're essentially copying pointers, not the data itself.

> I did a similar change at the end of last year and if I remember correctly the locking when getting the class slowed things down considerably. Think of people adding a bunch of tenants while doing other operations.

Adding tenants was already a safe operation. This PR primarily focuses on enhancing the safety of readers that handle query requests. The only writer involved is the schema updater.

`Furthermore, In this context it is crucial to prioritize correctness over performance. Ideally, if this issue was known beforehand, we should have addressed it earlier to avoid dealing with these kind of race issues`

Reply:
makes sense.

Would it be possible to do a chaos/benchmark to get an idea how much of a slowdown could happen?

Reply:
I conducted a load test using a cluster of 4 nodes with a replication factor of 2.
The test involved creating 600 tenants sequentially, ensuring that the 'readOnlyClass' was called 600 times. Here are the results: Without the fix, it took an average of 1m53s, and with the fix, it took 1m56s.

So the new version is slightly slower but safer then the current one